### PR TITLE
#311 implemented fragmented keys

### PR DESCRIPTION
**Pull request by vaukalak**
this PR solves #311

### CODE REVIEW

**Comment by vaukalak** [comment]
@rt2zz is there any chance, this will be reviewed?

File: src/createPersistor.js [hunk 1]
```diff
@@ -12,6 +12,7 @@ export default function createPersistor (store, config) {
   const transforms = config.transforms || []
   const debounce = config.debounce || false
   const keyPrefix = config.keyPrefix !== undefined ? config.keyPrefix : KEY_PREFIX
+  const createFragmentedKey = config._createFragmentedKey || defaultCreateFragmentedKey
```
Comment (Andarist) [inline src/createPersistor.js]: if u pass some option in the config, it for sure shouldnt be named like this `_createFragmentedKey` which indicates something private

**Comment by vaukalak** [comment]
@Andarist updated removed `_` in config keys?

**Comment by vaukalak** [comment]
@rt2zz @Andarist sorry for bumping, but do you have any questions / notes on that?

File: src/getStoredState.js [hunk 2]
```diff
@@ -55,14 +60,26 @@ export default function getStoredState (config, onComplete) {
...
-    if (whitelist && whitelist.indexOf(key) === -1) return false
-    if (blacklist.indexOf(key) !== -1) return false
+    if (whitelist && whitelist.map(mapKey).indexOf(key) === -1) return false
+    if (blacklist.map(mapKey).indexOf(key) !== -1) return false
     return true
   }
 
-  function createStorageKey (key) {
-    return `${keyPrefix}${key}`
+  function defaultFragmentKeyToReducerKey (key) {
```
Comment (Andarist) [inline src/getStoredState.js]: `defaultFragmentKeyToReducerKey` and `defaultCreateFragmentedKey` are the same and also they are just `identity` function 

`const identity = x => x`

**Comment by Andarist** [comment]
Ive added a single comment, but truth to be told im not that much involved in the project and Im not making any calls on new features etc, cannot review this properly as Im lacking the context for the feature and the actual codebase.

**Comment by rt2zz** [comment]
Apologize for the long delay, I have been working on getting v5 ready to go, and it is just about ready for primetime. I think v5 *may* solve the issue out of the box, or possibly some small modifications are required, lets discuss:

Please note, in v5 only **one** key/value is stored per persistReducer, so the solution is not directly analogous to the one in this PR.

If I understand the problem fully, the the basic goal is to have some portion of state partitioned based on some variable, in this case lets say its the `user` reducer, partitioned by `userId`. in v5 this can be modeled by only adding the `user` reducer once we know what the partition key is. so e.g.

```js
// event: user logs in
const persistConfig = { key: `user:${userId}`, storage }
const persistedUserReducer = persistReducer(persistConfig, userReducer)

store.replaceReducer(combineReducers({
  ...baseReducers,
  user: persistedUserReducer,
})
persistor.persist() // @NOTE this method does not yet exist.
```

If we want to go with this we need to add the persist method to the persistor which I have been thinking about doing regardless. 

Ergonomically I think the challenge here is that the partition key need to come from presumably other state. Which means we would always need to first persist the primary reducer, then persist again for the partitioned reducer. I do not see a way around this, but we might be able to make the external api simpler, e.g. avoid store.replaceReducer. That might look something like:
```js
let userReducer = persistReducer({ key, storage, persistKey: PERSIST_USER})

// ...

let persistor = persistStore(store, undefined, () => {
  let userId = store.getState().userId
  if (userId) {
    userReducer.replacePersistConfig({ key: `user:${userId}`, storage })
    persistor.persist(PERSIST_USER)
  }
})
```

I am not sure which approach I like more, the idea of adding a `replacePersistConfig` method to the reducer is not particularly appealing.

Thoughts?

**Comment by ScreamZ** [comment]
Any news on that people ? Now we are on V5 #637 along other need something like that ?

**Comment by vaukalak** [comment]
@ScreamZ I think this PR is good to close. @rt2zz can you please post duplicate your post in #637 . So we can discuss the desired api? thanks.

**Comment by Faolain** [comment]
@rt2zz any shot you can duplicate your post on #637 as vaukalak mentioned?

**Pull request closed without merging**